internal/bguploader.go

```diff
@@ -1,146 +1,171 @@
+	ctx context.Context
+	// cancelFunc signals internals to stop enqueuing more uploads
+	cancelFunc context.CancelFunc
 
-	preventWalOverwrite bool
+	// workerCount tracks number of concurrent uploaders. Limited to maxParallelWorkers
+	workerCount *semaphore.Weighted
+	// maxParallelWorkers usually defined by WALG_DOWNLOAD_CONCURRENCY
+	maxParallelWorkers int32
+
+	// numUploaded is tracked to control amount of work done in one cycle of archive_command
+	numUploaded int32
+	// maxNumUploaded controls the amount of work done in one cycle of archive_command
+	maxNumUploaded int32
+
+	// started tracks filenames of ongoing and complete uploads to avoid repeating work
+	started map[string]struct{}
 }
 
-func NewBgUploader(walFilePath string, maxParallelWorkers int32, uploader *WalUploader, preventWalOverwrite bool) *BgUploader {
-	started := make(map[string]interface{})
-	started[filepath.Base(walFilePath)+readySuffix] = walFilePath
+func NewBgUploader(walFilePath string, maxParallelWorkers int32, maxNumUploaded int32, uploader *WalUploader, preventWalOverwrite bool) *BgUploader {
+	started := make(map[string]struct{})
+	started[filepath.Base(walFilePath)+readySuffix] = struct{}{}
+	ctx, cancelFunc := context.WithCancel(context.Background())
 	return &BgUploader{
-		filepath.Dir(walFilePath),
-		0,
-		maxParallelWorkers,
-		sync.WaitGroup{},
-		started,
-		uploader,
-		0,
-		sync.Mutex{},
-		preventWalOverwrite,
+		dir:                 filepath.Dir(walFilePath),
+		uploader:            uploader,
+		preventWalOverwrite: preventWalOverwrite,
+
+		ctx:                ctx,
+		cancelFunc:         cancelFunc,
+		workerCount:        semaphore.NewWeighted(int64(maxParallelWorkers)),
+		maxParallelWorkers: maxParallelWorkers,
+		numUploaded:        0,
+		maxNumUploaded:     maxNumUploaded,
+		started:            started,
 	}
 }
 
 // Start up checking what's inside archive_status
-func (bgUploader *BgUploader) Start() {
-	if bgUploader.maxParallelWorkers < 1 {
-		return // Nothing to start
+func (b *BgUploader) Start() {
+	// Exit early if BgUploader is effectively disabled
+	if b.maxParallelWorkers < 1 || b.maxNumUploaded < 1 {
+		return
 	}
-	// This goroutine will spawn new if necessary
-	go bgUploader.scanOnce()
-}
 
-// Stop pipeline
-func (bgUploader *BgUploader) Stop() {
-	for atomic.LoadInt32(&bgUploader.parallelWorkers) != 0 {
-		time.Sleep(50 * time.Millisecond)
-	} // Wait until no one works
-
-	bgUploader.mutex.Lock()
-	// We have to do this under mutex to exclude interference with shouldKeepScanning() branch
-	atomic.StoreInt32(&bgUploader.maxParallelWorkers, 0) // stop new jobs
-	bgUploader.mutex.Unlock()
+	go b.scanFiles()
+}
 
-	bgUploader.running.Wait() // wait again for those who jumped to the closing door
+// Stop pipeline. Stop can be safely called concurrently and repeatedly.
+func (b *BgUploader) Stop() {
+	// Send signal to stop scanning for and uploading new files
+	b.cancelFunc()
```

Comment:
Maybe document why it's safe to call cancelFunc multiple times (https://golang.org/pkg/context/#CancelFunc):

>  A CancelFunc may be called by multiple goroutines simultaneously. After the first call, subsequent calls to a CancelFunc do nothing.